### DISCUSSION TITLE
Purchase Float

### DISCUSSION THREAD

**Discussion by justinkekeocha** [bavix/laravel-wallet#775] [General]
I can't find a way to make purchase for float amount in the documentation. when I try, it stores only Int amount in database

**Comment by rez1dent3** [comment]
Hello. Only the whole amount is always stored in the database. If you call the *float method, then the number calculated using the formula is saved to the database: amount * 10^decimal_places.

Let decimal_places =2. That is, you enter 100 into the depositFloat() method, but 100*10^2 (=10000) will be saved into the database.

Purchases operate under a strict contract. I did not want to divide them into two contracts (integers and fractions), but left only integers. The functionality, according to my analytics, is not very popular. It is used at first by startups, and later they implement their more complex functionality as part of the package.

I deliberately do not complicate the payment functionality with unnecessary functionality, because everyone needs different functionality and it is difficult to do something completely general.

If you have any ideas, I don't mind listening to them.

**Comment by justinkekeocha** [comment]
Thank you for your reply. Which float method are you referring to? I can't find float method for purchase. only for deposit, withdraw and the rest under the HasWalletFloat trait.

I am talking of purchases. Not regular depositFloat and Withdraw

**Reply by rez1dent3** [reply]
You understand that the cost will depend on the specific wallet and its decimal_places?

As I wrote above https://github.com/bavix/laravel-wallet/discussions/775#discussioncomment-7030271. No, the product contract obliges you to work exclusively with integers. I did not split it into two, so as not to overcomplicate the already complex logic of the package.

**Comment by justinkekeocha** [comment]
I also noticed that the paid method is deprecated in CartPay trait. Is there a new way to check if user have paid?
Or this is should be business logic too?

**Reply by rez1dent3** [reply]
```php
    /**
     * Checks acquired product your wallet.
     *
     * @deprecated The method is slow and will be removed in the future
     * @see PurchaseServiceInterface
     */
```
Yes, it is enough to use PurchaseServiceInterface. 

The code inside the paid method is up-to-date and can be used in your projects.

But in the standard implementation everything is done in a general form; for each project it is better to implement its own class from the PurchaseServiceInterface interface.

**Comment by justinkekeocha** [comment]
Check out this package [https://github.com/stephenjude/laravel-wallet](url). It has only one method that handles integer and float.
Since all numbers are converted to an Integer when being stored to database. whether it was used by depositFloat or just deposit.

If the deposit method handles float number by defualt, the purchase functionality wouldn't need two contracts.

If deposit, is written as :

` public function deposit(int|string $amount, ?array $meta = null, bool $confirmed = true): Transaction
    {  
        $math = app(MathServiceInterface::class);
        $decimalPlacesValue = app(CastServiceInterface::class)->getWallet($this)->decimal_places;
        $decimalPlaces = $math->powTen($decimalPlacesValue);
        $amount = $math->round($math->mul($amount, $decimalPlaces, $decimalPlacesValue));

        return app(AtomicServiceInterface::class)->block(
            $this,
            fn () => app(TransactionServiceInterface::class)
                ->makeOne($this, Transaction::TYPE_DEPOSIT, $amount, $meta, $confirmed)
        );
    }`
    
    
 I want to know the possible implications. Will values be lost as you previous showed me in your previous example in the other discussion.

  I have tried using depositFloat for integers, and it worked perfectly well. I don't see why this isn't the default mode for the deposit method.
  
  I am sorry I am asking too many questions and I appreciate the time you take to answer questions

**Reply by rez1dent3** [reply]
You don't implement N-decimal_places in a simple implementation. 

https://github.com/bavix/laravel-wallet/blob/a83b72acaf946557dd15d6df60fa2481d4d8c8fd/tests/Units/Domain/WalletFloatTest.php#L276-L281

https://github.com/bavix/laravel-wallet/blob/a83b72acaf946557dd15d6df60fa2481d4d8c8fd/tests/Units/Domain/WalletFloatTest.php#L292-L313

The issue is not the difficulty of leaving only one type, but the fact that the package will be 5 years old this year. We need to be conservative and support old projects. Also, do not forget that float methods in real time constantly recalculate the balance, which adds a large load. Conversion is not free.
